### ISSUE TITLE
How to integrate custom model, evaluation criteria and training tricks with OpenNMT?

### ISSUE THREAD

**Issue by PosoSAgapo** [OpenNMT/OpenNMT-py#2129]
OpenNMT is a fast and out-of-box tool and I enjoy it very much. However, how do we add user custom models, evaluation criteria, or training tricks based on the OpenNMT library?

For example, what should I do if I want to use my custom model rather than RNN/Transformer? Additionally, training tricks like learning rate scheduler, how do I use them in OpenNMT? Also, the evaluation criteria vary from dataset to dataset, how do I replace the official EM criteria with my custom criteria?

**Comment by francoishernandez** [comment]
> custom models

All the model related code lives in `onmt.encoder`, `onmt.decoder`, `onmt.models` and the model is actually glued together in `onmt.model_builder`.

> evaluation criteria

Are we talking about having different metrics for different training datasets? Or several valid sets ? Anyways, none of this is really handled right now. If you want some pointers on how to get it done you might want to give more details about exactly what you expect to achieve.

> learning rate scheduler

Quite a few possibilities are already implemented (https://opennmt.net/OpenNMT-py/options/train.html#Optimization-%20Rate).

If you want to customize this, have a look around here:
https://github.com/OpenNMT/OpenNMT-py/blob/2d2445b11091ff4ef837444f7db7a107ee454b09/onmt/utils/optimizers.py#L101-L124

**Comment by PosoSAgapo** [comment]
> Are we talking about having different metrics for different training datasets? Or several valid sets ? Anyways, none of this is really handled right now. If you want some pointers on how to get it done you might want to give more details about exactly what you expect to achieve.

Thank you for your reply, actually, I am talking about the different metrics in different datasets. 
OpenNMT is built for machine translation, but considering the NMT is a seq2seq problem, so the OpenNMT could also be considered as a Seq2Seq2 framework that expands to dataset beyond translation.  Take the semantic parsing for example, it can be thought of as a translate problem whether the model is asked to translate the human query into a machine-executable program, which the evaluation criteria might be the denotation accuracy ( which parses the output and verify where this out can be executed by some program, like testing whether this output can retrieve a  person in a database), however, this took different criteria this is not evaluated using Exact Match or perplexity. But, I do want to use the convenience of the OpenNMT library, so I have to further develop custom criteria. Therefore, I would like to know how to add custom criteria, the core is to take the output string and input it into certain custom criteria

**Comment by francoishernandez** [comment]
It's still unclear to me exactly what you want to do but it is probably similar to #1158 and #1641. There is no easy way to do this for now. It would probably be way easier for you to perform inference with checkpoints saved at various steps and apply your custom evaluation with custom scripts.

If you want to use different loss criteria for training, the story might be different though.

**Comment by PosoSAgapo** [comment]
> It's still unclear to me exactly what you want to do but it is probably similar to #1158 and #1641. There is no easy way to do this for now. It would probably be way easier for you to perform inference with checkpoints saved at various steps and apply your custom evaluation with custom scripts.
> 
> If you want to use different loss criteria for training, the story might be different though.

Ok, let me be more clear on this. Considering a semantic parsing situation:
Input: I want to find a Person called Jack.
Gold Program: find (Jack) in Table (Person)
Model Generation: find (Mary) in Table (Person)
The model is to take the input and generate the executable program. The criteria is to parse the program to see whether the Jack and Person are generated or not, where the Jack is not correctly generated in the example. To test this, I need to put my own criteria to evaluate the model rather than exatch match or perplexity

**Comment by PosoSAgapo** [comment]
> It's still unclear to me exactly what you want to do but it is probably similar to #1158 and #1641. There is no easy way to do this for now. It would probably be way easier for you to perform inference with checkpoints saved at various steps and apply your custom evaluation with custom scripts.
> 
> If you want to use different loss criteria for training, the story might be different though.

Is it hard to change the evaluation function? In my view, replacing the function that evaluates the exact match(EM) with my custom function should be able to do this, since calculating the exact match require the model to generate sequences. If the sequences are already generated, why it is hard to use a custom evaluation?

**Comment by francoishernandez** [comment]
>  Model Generation: find (Mary) in Table (Person)

> If the sequences are already generated, why it is hard to use a custom evaluation?

When training/validating, this does not exist. The model does not generate at this point.
It just learns to predict the next token, based on the previous ones.
Hence we have some accuracy (of next token) or ppl (based on all 'next-token' predictions).

Your best bet is to 1. use `translate` to generate predictions with checkpoints at various steps 2. evaluate these predictions with any metric/script you like.

If you want to contribute a better way of doing this on a valid set, feel free, but it may not be as trivial as you might think.

**Comment by PosoSAgapo** [comment]
> > Model Generation: find (Mary) in Table (Person)
> 
> > If the sequences are already generated, why it is hard to use a custom evaluation?
> 
> When training/validating, this does not exist. The model does not generate at this point. It just learns to predict the next token, based on the previous ones. Hence we have some accuracy (of next token) or ppl (based on all 'next-token' predictions).
> 
> Your best bet is to 1. use `translate` to generate predictions with checkpoints at various steps 2. evaluate these predictions with any metric/script you like.
> 
> If you want to contribute a better way of doing this on a valid set, feel free, but it may not be as trivial as you might think.

Ok, now I see why this is hard. Saving models is a quick solution, but this gets tricky when the model is large （pre-trained models）， since it might take a huge amount of disk space. However,  thank you for your reply anyway!

**Comment by PosoSAgapo** [comment]
I think I am able to contribute a way to do this in the valid set, but not this method currently cannot work in the Copy mechanism and also need to rewrite some functions in this OpenNMT library
